src/sqlfluff/rules/L003.py

```diff
@@ -1,15 +1,156 @@
 """Implementation of Rule L003."""
-from typing import List, Optional, Sequence, Tuple
+import dataclasses
+import itertools
+from typing import Dict, Iterable, List, Optional, Sequence, Tuple
 
 from sqlfluff.core.parser import WhitespaceSegment
-from sqlfluff.core.parser.segments import BaseSegment, RawSegment
+from sqlfluff.core.parser.segments import BaseSegment
 from sqlfluff.core.rules.functional import rsp, sp, Segments
 from sqlfluff.core.rules.base import BaseRule, LintResult, LintFix, RuleContext
 from sqlfluff.core.rules.doc_decorators import (
     document_fix_compatible,
     document_configuration,
 )
 from sqlfluff.core.templaters import TemplatedFile
+from sqlfluff.core.templaters.base import RawFileSlice
+
+
+class _DictLikeDataCls:
+    def __getitem__(self, item):
+        return getattr(self, item)
+
+    def __setitem__(self, key, item):
+        return setattr(self, key, item)
+
+    def items(self):
+        for field_el in dataclasses.fields(self):
+            key = field_el.name
+            yield key, getattr(self, key)
+
+
+@dataclasses.dataclass
+class _LineSummary(_DictLikeDataCls):
+    """A dataobject to represent a line.
+
+    Doubles as a memory object (holding state)
+    that can be converted to a fresh summary instance
+    """
+
+    line_no: int = 0
+    templated_line: Optional[int] = None
+    line_buffer: List[BaseSegment] = dataclasses.field(default_factory=list)
+    ignoreable_reference: List[bool] = dataclasses.field(default_factory=list)
+    indent_buffer: List[BaseSegment] = dataclasses.field(default_factory=list)
+    indent_size: int = 1
+    indent_balance: int = 0
+    hanging_indent: Optional[int] = None
+    clean_indent: bool = True
+    template_content: str = ""
+    templated_line_type: Optional[str] = None
+    line_anchor: Optional[BaseSegment] = None
+
+    # These are only required for carrying state between lines
+    anchor_indent_balance: int = 0
+    in_indent: bool = True
+    line_indent_stack: List[int] = dataclasses.field(default_factory=list)
+    hanger_pos: Optional[int] = None
+
+    def strip_buffers(self) -> dict:
+        """Strip a line dict of buffers for logging."""
+        keys_to_strip = ("line_buffer", "indent_buffer", "ignoreable_reference")
+        print_dict: Dict = {
+            key: value for key, value in self.items() if key not in keys_to_strip
+        }
+        print_dict["raw"] = "".join(el.raw for el in self.line_buffer)
+        return print_dict
+
+    def trigger_el(self):
+        """Get the pivot element."""
```

Comment:
Can you add some detail about what you mean by "pivot element"? A brief SQL example could be helpful for context.

Reply:
Im still figuring out, its essentially the "tigger el". L003 was using a token to mark where to stop. This caused all previous line to be fully computed and the most recent line only to be computed up to the first non indentation token.

Id like to capture this logic such that each line knows about / holds enough info on both states. (allowing us to do a single pass) and also to consider senarios  where the whole of the current line is required for computation

eg my `\n))))` example where the bracket legitimately close off a bunch of indents